File: 94_binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/binary-tree-maximum-path-sum
@Language: Java
@Datetime: 16-05-19 08:35
*/

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: An integer.
     */
    private class returnType{
        int single;
        int maxpath;
        public returnType (int single, int maxpath) {
            this.single = single;
            this.maxpath = maxpath;
        }
    }
    private returnType helper(TreeNode root) {
        if (root == null) {
            return new returnType(0, Integer.MIN_VALUE);
        }
        returnType left = helper(root.left);
        returnType right = helper(root.right);
        int single = Math.max(0, Math.max(left.single + root.val, right.single + root.val));
        int maxpath = Math.max(left.single + right.single + root.val, Math.max(left.maxpath, right.maxpath));
        return new returnType(single, maxpath);
    }
    public int maxPathSum(TreeNode root) {
        // write your code here
        return helper(root).maxpath;
    }
}

File: 94_binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/binary-tree-maximum-path-sum
@Language: Python
@Datetime: 16-05-19 03:43
'''

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        # write your code here
        maxsum, _ = self.helper(root)
        return maxsum
        
    def helper(self, root):
        if root is None:
            return -sys.maxint, 0
        left = self.helper(root.left)
        right = self.helper(root.right)
        single = max(left[1]+root.val, right[1]+root.val, 0)
        maxpath = max(left[0], right[0], left[1] + right[1] + root.val)
        return maxpath, single
            
        